### PR TITLE
Sbiswas/identify kml features

### PR DESCRIPTION
**Pull request by sbiswas96**
Please review this sample that loads a KML layer. When the user clicks on one of the KML features, a callout will display the feature information

### CODE REVIEW

File: ArcGISRuntimeSDKQt_CppSamples/Layers/IdentifyKmlFeatures/IdentifyKmlFeatures.cpp [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+namespace  {
+const QUrl datasetUrl("https://www.wpc.ncep.noaa.gov/kml/noaa_chart/WPC_Day1_SigWx.kml");
+}
+
+IdentifyKmlFeatures::IdentifyKmlFeatures(QObject* parent /* = nullptr */):
+  QObject(parent),
+  m_map(new Map(Basemap::darkGrayCanvasVector(this), this))
+{
+
```
Comment (lsmallwood) [inline ArcGISRuntimeSDKQt_CppSamples/Layers/IdentifyKmlFeatures/IdentifyKmlFeatures.cpp]: remove empty line

File: ArcGISRuntimeSDKQt_CppSamples/Layers/IdentifyKmlFeatures/IdentifyKmlFeatures.cpp [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  if (!mapView || mapView == m_mapView)
+    return;
+
+  m_mapView = mapView;
+  m_mapView->setMap(m_map);
+
+  // start zoomed in over the US
+  m_mapView->setViewpointGeometry(Envelope(-19195297.778679, 512343.939994, -3620418.579987, 8658913.035426, SpatialReference::webMercator()));
+
+  KmlDataset* forecastDataset = new KmlDataset(datasetUrl, this);
```
Comment (lsmallwood) [inline ArcGISRuntimeSDKQt_CppSamples/Layers/IdentifyKmlFeatures/IdentifyKmlFeatures.cpp]: I think the KmlDataset can all be moved to the ctor

File: ArcGISRuntimeSDKQt_CppSamples/Layers/IdentifyKmlFeatures/IdentifyKmlFeatures.cpp [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    {
+      if (KmlPlacemark* placemark = dynamic_cast<KmlPlacemark*>(geoElement))
+      {
+        m_calloutText = placemark->balloonContent();
+        m_mapView->calloutData()->setLocation(m_clickedPoint);
+        m_mapView->calloutData()->setVisible(true);
+
+        emit calloutDataChanged();
+        emit calloutTextChanged();
+      }
```
Comment (lsmallwood) [inline ArcGISRuntimeSDKQt_CppSamples/Layers/IdentifyKmlFeatures/IdentifyKmlFeatures.cpp]: looks like you need to `break` out when you have found one? Or `return`

File: ArcGISRuntimeSDKQt_CppSamples/Layers/IdentifyKmlFeatures/IdentifyKmlFeatures.qml [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+// limitations under the License.
+// [Legal]
+
+import QtQuick 2.6
+import QtQuick.Controls 2.2
+import Esri.Samples 1.0
+import Esri.ArcGISRuntime.Toolkit.Controls 100.6 // needed to use Callout in QML
+
+Item {
+
```
Comment (lsmallwood) [inline ArcGISRuntimeSDKQt_CppSamples/Layers/IdentifyKmlFeatures/IdentifyKmlFeatures.qml]: remove empty line

File: ArcGISRuntimeSDKQt_CppSamples/Layers/IdentifyKmlFeatures/README.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+## Use case
+
+A user may wish to select a KML feature to view relevant information about it.
+
+## How to use the sample
+
+Click a feature to identify it. Feature information will be displayed in a callout.
+
+Note: the KML layer used in this sample contains a screen overlay. The screen overlay contains a legend and the logos for NOAA and the NWS. You can't identify the screen overlay.
```
Comment (lsmallwood) [inline ArcGISRuntimeSDKQt_CppSamples/Layers/IdentifyKmlFeatures/README.md]: should probably expand what the NOAA and NWS acronyms are (I have no idea)

File: ArcGISRuntimeSDKQt_QMLSamples/Layers/IdentifyKmlFeatures/IdentifyKmlFeatures.qml [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+                id: forecastLayer
+                KmlDataset {
+                    url: layerUrl
+                }
+            }
+        }
+
+        onMouseClicked: {
+            clickedPoint = screenToLocation(mouse.x, mouse.y);
+            identifyLayer(forecastLayer, mouse.x, mouse.y, 15, false, 1);
```
Comment (lsmallwood) [inline ArcGISRuntimeSDKQt_QMLSamples/Layers/IdentifyKmlFeatures/IdentifyKmlFeatures.qml]: maybe you should check there is not a task in flight before kicking off a new one?

File: ArcGISRuntimeSDKQt_QMLSamples/Layers/IdentifyKmlFeatures/IdentifyKmlFeatures.qml [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        }
+
+        onIdentifyLayerStatusChanged: {
+            if (identifyLayerStatus === Enums.TaskStatusCompleted) {
+                calloutText = identifyLayerResult.geoElements[0].balloonContent;
```
Comment (lsmallwood) [inline ArcGISRuntimeSDKQt_QMLSamples/Layers/IdentifyKmlFeatures/IdentifyKmlFeatures.qml]: you should check that `geoElements` is not empty

File: ArcGISRuntimeSDKQt_CppSamples/Layers/IdentifyKmlFeatures/IdentifyKmlFeatures.qml [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// [Legal]
+
+import QtQuick 2.6
+import QtQuick.Controls 2.2
+import Esri.Samples 1.0
+import Esri.ArcGISRuntime.Toolkit.Controls 100.6 // needed to use Callout in QML
```
Comment (ldanzinger) [inline ArcGISRuntimeSDKQt_CppSamples/Layers/IdentifyKmlFeatures/IdentifyKmlFeatures.qml]:
```suggestion
import Esri.ArcGISRuntime.Toolkit.Controls 100.8 // needed to use Callout in QML
```

File: ArcGISRuntimeSDKQt_QMLSamples/Layers/IdentifyKmlFeatures/IdentifyKmlFeatures.qml [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+
+Rectangle {
+    id: rootRectangle
+    clip: true
+    width: 800
+    height: 600
+
+    property Point clickedPoint: null
+    property string calloutText: ""
+    property url layerUrl: "https://www.wpc.ncep.noaa.gov/kml/noaa_chart/WPC_Day1_SigWx.kml"
```
Comment (ldanzinger) [inline ArcGISRuntimeSDKQt_QMLSamples/Layers/IdentifyKmlFeatures/IdentifyKmlFeatures.qml]:
```suggestion
    readonly property url layerUrl: "https://www.wpc.ncep.noaa.gov/kml/noaa_chart/WPC_Day1_SigWx.kml"
```

File: ArcGISRuntimeSDKQt_QMLSamples/Layers/IdentifyKmlFeatures/IdentifyKmlFeatures.qml [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        onMouseClicked: {
+            clickedPoint = screenToLocation(mouse.x, mouse.y);
+            identifyLayer(forecastLayer, mouse.x, mouse.y, 15, false, 1);
+        }
+
+        onIdentifyLayerStatusChanged: {
+            if (identifyLayerStatus === Enums.TaskStatusCompleted) {
+                calloutText = identifyLayerResult.geoElements[0].balloonContent;
+                callout.calloutData.location = clickedPoint;
+                callout.calloutContent = customComponent;
```
Comment (ldanzinger) [inline ArcGISRuntimeSDKQt_QMLSamples/Layers/IdentifyKmlFeatures/IdentifyKmlFeatures.qml]: can this just be done using property binding in the declaration of callout up between lines 35-42?

File: ArcGISRuntimeSDKQt_CppSamples/Layers/IdentifyKmlFeatures/README.md (hunk 5 above)
Comment (sbiswas96) [reply ArcGISRuntimeSDKQt_CppSamples/Layers/IdentifyKmlFeatures/README.md]: @ldanzinger thoughts on adding more explanation to the README?

Comment (ldanzinger) [reply ArcGISRuntimeSDKQt_CppSamples/Layers/IdentifyKmlFeatures/README.md]: @sbiswas96 in cases like this, I'd suggest updating it here first, then updating the common design. We will have a process to bubble out these changes elsewhere later

File: ArcGISRuntimeSDKQt_QMLSamples/Layers/IdentifyKmlFeatures/IdentifyKmlFeatures.qml (hunk 10 above)
Comment (ldanzinger) [reply ArcGISRuntimeSDKQt_QMLSamples/Layers/IdentifyKmlFeatures/IdentifyKmlFeatures.qml]: I think this cleans it up. Make sure to test out that if you click on several features, it continues to update/notify properly. I could see the binding perhaps not getting notified and then the same content would be shown for every click

**Pull request merged**